lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -774,8 +824,83 @@ void FIRRTLTypesLowering::getAllBundleLowerings(
       .Default([&](auto) {});
 }
 
-Identifier FIRRTLTypesLowering::getArgAttrName(unsigned newArgNumber) {
+Identifier TypeLoweringVisitor::getArgAttrName(unsigned newArgNumber) {
   SmallString<16> argNameBuffer;
   mlir::impl::getArgAttrName(newArgNumber, argNameBuffer);
-  return Identifier::get(argNameBuffer, &getContext());
+  return Identifier::get(argNameBuffer, context);
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
+  void runOnOperation() override;
+
+private:
+  void runAsync();
+  void runSync();
+  // A list of visitors, one for each thread.  These are cached by the pass.
+  std::vector<TypeLoweringVisitor> visitors;
+};
+} // end anonymous namespace
+
+void LowerTypesPass::runAsync() {
+  // Create a TypeLoweringVisitor for each thread to use.
+  size_t numThreads = llvm::hardware_concurrency().compute_thread_count();
+  visitors.resize(numThreads, &getContext());
+
+  // Used to divide up the work, first come first serve.
+  std::atomic<unsigned> opIt(0);
+
+  // Process the operations in parallel, with one runner for each hardware
+  // thread.  Each thread will get an operation index to process, and walk the
+  // list of operations until it finds that operation. It will repeat this until
+  // there are no more operations in the list.
+  mlir::ParallelDiagnosticHandler diagHandler(&getContext());
+  auto &ops = getOperation().getBody()->getOperations();
+  auto perThread = [&](TypeLoweringVisitor &visitor) {
+    // This is the operation index that this thread should process.
+    auto nextId = opIt++;
+    // This is the current operation index.
+    unsigned curId = 0;
+    for (auto &op : ops) {
+      // Skip forward until we are at operation this thread should process.
+      if (curId++ != nextId) {
+        continue;
+      }
+
+      // Notify the handler the op index and then perform lowering.
+      diagHandler.setOrderIDForThread(curId);
+      visitor.lowerModule(&op);
+      diagHandler.eraseOrderIDForThread();
+
+      // Get the next op to process.
+      nextId = opIt++;
+    }
+  };
+  llvm::parallelForEach(visitors.begin(), visitors.end(), perThread);
```

Comment:
I'd change this to work over `ops.begin(), ops.end()`, this allows foreach to clump based on the concurrency level present.

Reply:
This would be simpler, the truth behind this annoyingly complicated implementation is that I was being conservative about changing the performance characteristics, so I was following what the PassManager did.

I believe that parallelFor will `TaskGroup.spawn()` for each element in the range (up to 1024 tasks), it doesn't currently chunk the tasks based on the hardware. There is a comment in Parallel.h that made me nervous:
```
// TaskGroup has a relatively high overhead, so we want to reduce
// the number of spawn() calls. We'll create up to 1024 tasks here.
// (Note that 1024 is an arbitrary number. This code probably needs
// improving to take the number of available cores into account.)
```
I think under the hood there is a proper thread pool, this just creates more tasks than necessary.  I think it makes more sense to switch to use parallelFor more normally, and fix upstream if there is an issue.

And a side note, the implementation calls std::distance too many times, which is bad for our linked list.  This is just a note for me to fix it upstream.

Reply:
I tried switching strategies, its a lot simpler but still not as simple as I think we were hoping.  New comment chain further down.

Reply:
Does the foreach do stealing?  I'd hate to clump based on cores or some static number if the scheduler didn't at least do stealing (which is non-trivial to do efficiently, you need runtime cache topologies, core-bound threads, etc, but even a bounded simple version should avoid the worst pathologies of clumping tasks togeather).